### PR TITLE
Adds webscraper package

### PR DESCRIPTION
**Pull request by jar-o**
the rationale behind a generalized webscraper package is that if we have to scrape price for one model, we will likely have to do it for another. `webscraper` provides

1. simple semantics for scraping (the underlying `goquery` package is not specifically designed for scraping),
2. generalizes those scraping semantics for use in other price models,
3. preloading of docs for optimization of multiple scrapes,
4. compatibility with the gofer request pool, and
5. easy debug flagging for exposure of the underlying html document

### CODE REVIEW

File: pkg/webscraper/scraper.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	if o.debug {
+		os.Stderr.WriteString(string(b))
+	}
+	doc, err := goquery.NewDocumentFromReader(bytes.NewReader(b))
+	if err != nil {
+		return nil, err
+	}
+	o.preloadedDoc = doc
+	return o, nil
+}
```
Comment (jar-o) [inline pkg/webscraper/scraper.go:71]: note, this func exists specifically to take advantage of the request pool that Gofer uses...

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -103,6 +103,7 @@ jobs:
         uses: actions/checkout@v2
       - name: Download vendors
         run: go mod vendor
+        run: go mod tidy
```
[the remark this answers is not in the document]
Comment (jar-o) [reply .github/workflows/test.yml:106]: it was for github actions to pass 🤷

**Review by teghnet — APPROVED** [review]

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jar-o) [reply .github/workflows/test.yml]: welp

I had some go.mod conflicts this morning to resolve and so thought I'd run tidy locally and remove this... no luck. the ghost e2e tests fail

https://github.com/chronicleprotocol/oracle-suite/actions/runs/3960285596/jobs/6784229606

File: pkg/webscraper/scraper.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/PuerkitoBio/goquery"
+)
+
+type Scraper struct {
+	preloadedDoc *goquery.Document
+	debug        bool
+}
+
+type Element struct {
+	Text  string
```
Comment (jar-o) [inline pkg/webscraper/scraper.go:38]: removed the exposure of `goquery.Selection` in return values

File: .github/workflows/test.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jar-o) [reply .github/workflows/test.yml]: oh derp, my bad. I was running tidy from the wrong level heh. this is fixed

File: pkg/webscraper/scraper.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+				Text:  s.Text(),
+				Index: i,
+			})
+		},
+	)
+	return nil
+}
+
+func (o *Scraper) loadDoc(url string) (*goquery.Document, error) {
+	res, err := http.Get(url) //nolint:gosec
```
Comment (MDobak) [inline pkg/webscraper/scraper.go:105]: http.Get should use context.

File: pkg/webscraper/scraper.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		return nil, fmt.Errorf("status code error: %d %s", res.StatusCode, res.Status)
+	}
+
+	if o.debug {
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		res.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+		os.Stderr.WriteString(string(body))
```
Comment (MDobak) [inline pkg/webscraper/scraper.go:121]: Instead of the `debug` flag our logger can be used.

File: pkg/webscraper/scraper.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	doc, err := goquery.NewDocumentFromReader(res.Body)
```
Comment (MDobak) [inline pkg/webscraper/scraper.go:124]: Var assignment can be replaced with `return`.

**Comment by MDobak** [comment]
A few minor things, other than that lgtm.

File: pkg/webscraper/scraper.go (hunk 6 above)
Comment (jar-o) [reply pkg/webscraper/scraper.go]: ah yup, fixed

File: pkg/webscraper/scraper.go (hunk 5 above)
Comment (jar-o) [reply pkg/webscraper/scraper.go]: oh, rt. good call. fixed

File: pkg/webscraper/scraper.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package webscraper
```
[the remark this answers is not in the document]
Comment (jar-o) [reply pkg/webscraper/scraper.go:16]: kk 👍

**Pull request merged**